accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -207,26 +215,30 @@ else if (failure != null)
 
     private void ack()
     {
+        // wait for -1 because to ensure the setup phase has also completed. Setup calls ack in its callback
+        // and prevents races where we respond before dispatching all the required reads (if the reads are
+        // completing faster than the reads can be setup on all required shards)
         if (-1 == --waitingOnCount)
             node.reply(replyTo, replyContext, new ReadOk(data));
     }
 
     private synchronized void readComplete(CommandStore commandStore, Data result)
     {
-        Preconditions.checkState(waitingOn.get(commandStore.id()));
+        Invariants.checkState(waitingOn.get(commandStore.id()));
         logger.trace("{}: read completed on {}", txnId, commandStore);
         if (result != null)
             data = data == null ? result : data.merge(result);
 
+        Invariants.checkState(waitingOn.get(commandStore.id()));
```

Comment:
can we get a error msg?  if this fails in prod it can be hard to know was this pre or post `data.merge`, though I don't understand the case this happens.  We only remove in this method and we only mutate the BitSet around `synchronized` blocks, so its not clear how this condition could fail after the first was success